WordPress/openverse-catalog
# Add iNaturalist.org metadata

## Fixes
Fixes #439  by @sarayourfriend 

## Description
The [iNaturalist API](https://api.inaturalist.org/v1/docs/) is not intended for data scraping, but there is a [full dump](https://github.com/inaturalist/inaturalist-open-data/tree/documentation/Metadata) intended for sharing on S3. Those files cover over 120 million images, with information about the person who shared the photo and the species in the photo represented across four normalized tables. 

This PR loads *all* of the inaturalist metadata at once. The source files do not include any date or time fields to support differentiating among duplicate IDs or performing incremental updates. However, image files are in the same S3 bucket, prefixed by the photo_id in the metadata files. So, future dagruns could use [s3Hook.list_keys](https://airflow.apache.org/docs/apache-airflow-providers-amazon/3.3.0/_api/airflow/providers/amazon/aws/hooks/s3/index.html#airflow.providers.amazon.aws.hooks.s3.S3Hook.list_keys) to pull photo ID prefixes with files that have been updated since the [prev_start_date_success](https://airflow.apache.org/docs/apache-airflow/stable/templates-ref.html) and process only those photo IDs each month.

## Testing Instructions
Check out the branch. And then this PR includes changes to the docker dev setup, so it's a good idea to use `just down -v` and `just recreate` when you're testing. 

`just test` will cover basic tests of the provider data ingestion class functions, but not the DAG steps that make the metadata available .

There are very small subsets of the real files in /tests/s3-data/inaturalist-open-data, that will be loaded to the dev environment equivalent of s3, so that you can go kick off the inaturalist_workflow dag in Airflow, and inspect the logs and the resulting tsv file and the `image` table in the db. The Airflow log will give you the exact file name, and you can see it by running `just shell` and `more /var/workflow_output/inaturalist_image_v001_20220709220913.tsv` (or whatever the file is called). Run the dag a second time to confirm that the skips work when there is no new data.

I tried putting the real fullsize inaturalist files in /tests/s3-data/inaturalist-open-data, running `just down -v` and `just recreate`, and waiting a little bit for minio to finish loading the files. I've not been able to process the full dataset in my dev environment, but for now I am assuming that it is due to limitations of the environment. 

## Checklist



[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
